pkg/pipeline/pipeline.go

```diff
@@ -21,58 +21,53 @@ import (
 	"fmt"
 
 	"github.com/heptiolabs/healthcheck"
+	"github.com/mariomac/pipes/pkg/graph"
+	"github.com/mariomac/pipes/pkg/node"
 	"github.com/netobserv/flowlogs-pipeline/pkg/config"
-	"github.com/netobserv/gopipes/pkg/node"
-	log "github.com/sirupsen/logrus"
+	"github.com/netobserv/flowlogs-pipeline/pkg/pipeline/decode"
+	"github.com/netobserv/flowlogs-pipeline/pkg/pipeline/ingest"
 )
 
-// interface definitions of pipeline components
-const (
-	StageIngest    = "ingest"
-	StageDecode    = "decode"
-	StageTransform = "transform"
-	StageExtract   = "extract"
-	StageEncode    = "encode"
-	StageWrite     = "write"
-)
+//TODO: can go into config.Options
+const channelBufferLen = 20
 
 // Pipeline manager
 type Pipeline struct {
-	startNodes    []*node.Init
-	terminalNodes []*node.Terminal
-	IsRunning     bool
-	// TODO: this field is only used for test verification. We should rewrite the build process
-	// to be able to remove it from here
-	pipelineStages []*pipelineEntry
+	IsRunning bool
+	graph     graph.Graph
 }
 
 // NewPipeline defines the pipeline elements
-func NewPipeline() (*Pipeline, error) {
-	log.Debugf("entering NewPipeline")
+func NewPipeline(options config.Options) (*Pipeline, error) {
+	builder := graph.NewBuilder(node.ChannelBufferLen(channelBufferLen))
+
+	graph.RegisterStart(builder, ingest.FileProvider)
+	graph.RegisterStart(builder, ingest.CollectorProvider)
+	graph.RegisterStart(builder, ingest.GRPCProtobufProvider)
+	graph.RegisterStart(builder, ingest.IngestKafkaProvider)
 
-	stages := config.PipeLine
-	log.Debugf("stages = %v ", stages)
-	configParams := config.Parameters
-	log.Debugf("configParams = %v ", configParams)
+	graph.RegisterMiddle(builder, decode.AwsProvider)
+	graph.RegisterMiddle(builder, decode.JSONProvider)
+	graph.RegisterMiddle(builder, decode.ProtobufProvider)
 
-	build := newBuilder(configParams, stages)
-	if err := build.readStages(); err != nil {
+	/** etc **/
+
+	graph, err := builder.Build(options.PipeLine)
```

Comment:
the whole pipeline builder is replaced by this... we should add more `RegisterMiddle` and `RegisterTerminal` for the rest of stages.

The encoder/decoders could be replaced by `RegisterCodec` invocations, so you don't need to explicitly instantiate them in the configuration.

The `builder.Build` method magically instantiates and connects all the nodes according to the passed pipeline configuration.